pkgs/development/python-modules/mip/default.nix

```diff
@@ -0,0 +1,90 @@
+{ lib
+, stdenv
+, buildPythonPackage
+, cbc
+, cffi
+, dos2unix
+, fetchPypi
+, matplotlib
+, networkx
+, numpy
+, pytestCheckHook
+, pythonOlder
+, gurobi
+, gurobipy
+# Enable support for the commercial Gurobi solver (requires a license)
+, gurobiSupport ? false
+# If Gurobi has already been installed outside of the Nix store, specify its
+# installation directory here
+, gurobiHome ? null
+}:
+
+buildPythonPackage rec {
+  pname = "mip";
+  version = "1.14.1";
+
+  disabled = pythonOlder "3.7";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "6efa66e6f529d797dbbff4b0d4bc7bd22840ed51ec494cf0173a050c6f8ec353";
+  };
+
+  checkInputs = [ matplotlib networkx numpy pytestCheckHook ];
+  nativeBuildInputs = [ dos2unix ];
+  propagatedBuildInputs = [
+    cffi
+    cbc
+  ] ++ lib.optionals gurobiSupport ([
+    gurobipy
+  ] ++ lib.optional (builtins.isNull gurobiHome) gurobi);
+
+  # Source files have CRLF terminators, which make patch error out when supplied
+  # with diffs made on *nix machines
+  prePatch = ''
+    find . -type f -exec ${dos2unix}/bin/dos2unix {} \;
+  '';
+
+  patches = [
+    # Allow cffi versions with a different patch level to be used
+    ./cffi-version.patch
+    # Some tests try to be smart and dynamically construct a path to their test
+    # inputs. Unfortunately, since the test phase is run after installation,
+    # those paths point to the Nix store, which no longer contains the test
+    # data. This patch hardcodes the data path to point to the source directory.
+    ./test-data-path.patch
+  ];
+
+  # buildPythonPackage invokes setup.py directly, but MIP does not provide one
+  postPatch = ''
+    cat > setup.py << EOF
```

Comment:
What is the setup process then? Is it a good idea to add an ad-hoc setup.py?

Reply:
Unfortunately, it is our buildPythonPackage to be at fault, here.

The world is moving on with [PEP 517](https://peps.python.org/pep-0517/), and distutils [has been deprecating](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#setuppy-discouraged) direct invocation of setup.py. In order to support older workflows, the current recommendation is to [create a stub setup.py](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#setup-py) containing only a call to `setup()`, which will then cause the build system to parse `pyproject.toml`.

This will no longer be necessary once our Python infrastructure will correctly support PEP 517-style builds, where [build](https://pypa-build.readthedocs.io/en/latest/) is invoked as a generic frontend driver, instead.

Reply:
Have you tried: `format = "pyproject"`? see https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonpackage-parameters-buildpythonpackage-parameters

Reply:
You are right, it seems to be working.
For some reason I ruled it out because it was not invoking the builder correctly, as it is relying on a direct invocation of `wheel`, instead. I thought that was worse than a direct use of `setup.py`.

Sorry for the misjudgement, correcting now.